### PR TITLE
start historical detector

### PR DESCRIPTION
**Pull request by ylwu-amzn**
*Issue #, if available:*

## Description of changes:
If a detector has detection date range, we call it "historical detector".

Use the same start detector API to start both realtime and historical detector. 
1. We don't change realtime detector execution logic.
2. For historical detector, we will start an AD task and return the task id to client, then dispatch the task to a node with least load. The AD task will be split into smaller pieces, one piece has 1000 detection intervals by default, user can tune the piece size dynamically. AD task runner will execute the pieces sequentially. 

### AD task states
ad_workbench_task_lifecycle (2) (1)

## Test
1. `./gradlew build`
2. `./gradlew integTest -PnumNodes=3`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/ADTaskCancelledException.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/common/exception/ADTaskCancelledException.java:2]: Not sure why, `./gradlew spotlessApply` will reset copyright year as 2020. If change to 2021, `./gradlew build` will fail. Will investigate more.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 2]
```diff
@@ -699,4 +699,63 @@ public int getShingleSize(String detectorId) {
             return -1;
         }
     }
+
+    public void getFeatureDataPoints(
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:703]: maybe name it `getFeatureDataPointsByBatch`, `getFeatureDataPoints` sounds like it can be used for any general case.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskManager.java [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+            } catch (Exception e) {
+                String message = "Failed to start anomaly detector";
+                logger.error(message, e);
+                listener.onFailure(new ElasticsearchStatusException(message, RestStatus.INTERNAL_SERVER_ERROR));
+            }
+        }, exception -> listener.onFailure(exception)));
+    }
+
+    private String validateDetector(AnomalyDetector detector) {
+        if (detector.getFeatureAttributes().size() == 0) {
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskManager.java:179]: is it possible that `getFeatureAttributes()` is null?

File: src/test/java/com/amazon/opendistroforelasticsearch/ad/plugin/MockReindexPlugin.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import com.google.common.collect.ImmutableList;
+
+public class MockReindexPlugin extends Plugin implements ActionPlugin {
+
+    @Override
+    public List<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
+        return Arrays
+            .asList(
+                new ActionHandler<>(UpdateByQueryAction.INSTANCE, MockTransportUpdateByQueryAction.class),
+                new ActionHandler<>(DeleteByQueryAction.INSTANCE, MockTransportDeleteByQueryAction.class)
```
Comment (yizheliu-amazon) [inline src/test/java/com/amazon/opendistroforelasticsearch/ad/plugin/MockReindexPlugin.java:62]: question: how is `MockTransportUpdateByQueryAction ` and `MockTransportDeleteByQueryAction ` being used in test?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+                        .getStatsMap()
+                        .get(AD_EXECUTING_BATCH_TASK_COUNT.getName()))
+                            .compareTo((Long) r2.getStatsMap().get(AD_EXECUTING_BATCH_TASK_COUNT.getName()))
+                )
+                .collect(Collectors.toList());
+
+            if (candidateNodeResponse.size() == 1) {
+                listener.onResponse(candidateNodeResponse.get(0).getNode());
+            } else {
+                // if multiple nodes have same running task count, choose the one with least JVM heap usage.
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:262]: JVM heap usage is not so stable, it may change quickly. Some node may have lower JVM heap usage, but it may increase rapidly and historical detector will run for a relatively longer time (10 minutes SLA for 1 month data). It's hard to predict the future JVM heap usage. If a node's JVM heap usage not exceed our limit(85%), then we consider it can run new AD task. We will not dispatch AD task to node which has higher heap usage than 85%. So we'd better distribute AD tasks evenly on data nodes.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+
+        client.execute(ADStatsNodesAction.INSTANCE, adStatsRequest, ActionListener.wrap(adStatsResponse -> {
+            List<ADStatsNodeResponse> candidateNodeResponse = adStatsResponse
+                .getNodes()
+                .stream()
+                .filter(stat -> (long) stat.getStatsMap().get(JVM_HEAP_USAGE.getName()) < DEFAULT_JVM_HEAP_USAGE_THRESHOLD)
+                .collect(Collectors.toList());
+
+            if (candidateNodeResponse.size() == 0) {
+                String errorMessage = "All nodes' memory usage exceeds limitation. No eligible node to run detector "
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:232]: Good point, will add

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/indices/AnomalyDetectionIndices.java [hunk 7]
```diff
@@ -376,26 +376,33 @@ public void initAnomalyResultIndexDirectly(ActionListener<CreateIndexResponse> a
...
      * @param actionListener action called after create index
-     * @throws IOException IOException from {@link AnomalyDetectionIndices#getDetectorStateMappings}
      */
-    public void initDetectorStateIndex(ActionListener<CreateIndexResponse> actionListener) throws IOException {
-        CreateIndexRequest request = new CreateIndexRequest(DetectorInternalState.DETECTOR_STATE_INDEX)
-            .mapping(AnomalyDetector.TYPE, getDetectorStateMappings(), XContentType.JSON);
-        adminClient.indices().create(request, markMappingUpToDate(ADIndex.STATE, actionListener));
+    public void initDetectionStateIndex(ActionListener<CreateIndexResponse> actionListener) {
+        try {
+            CreateIndexRequest request = new CreateIndexRequest(DetectorInternalState.DETECTOR_STATE_INDEX)
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/indices/AnomalyDetectionIndices.java:399]: `DetectorInternalState ` is only for detector's internal state, I think it's ok to keep as is. As `ADTask` and `DetectorInternalState` both defines constants for detection index, will move it to `CommonName`.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 8]
```diff
@@ -699,4 +699,63 @@ public int getShingleSize(String detectorId) {
...
+            .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+        List<Entry<Long, Long>> missingRanges = getMissingRangesInShingle(detector, featuresMap, endTime);
+        missingRanges.stream().forEach(r -> {
+            if (dataPoints.containsKey(r.getKey())) {
+                featuresMap.put(r.getValue(), new SimpleImmutableEntry<>(r.getValue(), dataPoints.get(r.getKey())));
+            }
+        });
+        shingle.clear();
+
+        shingle.clear();
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:736]: good catch, will remove one.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 2 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:703]: sounds a better name, will change

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskManager.java (hunk 3 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskManager.java:179]: Good catch, it's possible. Will add null check.

File: src/test/java/com/amazon/opendistroforelasticsearch/ad/plugin/MockReindexPlugin.java (hunk 4 above)
Comment (ylwu-amzn) [reply src/test/java/com/amazon/opendistroforelasticsearch/ad/plugin/MockReindexPlugin.java:62]: `MockReindexPlugin.java` will be loaded by `AnomalyDetectorJobTransportActionTests#getMockPlugins`, then these two transport actions will be registered into test cluster. When we use `UpdateByQueryAction` and `DeleteByQueryAction` in `ADTaskManager`, these two transport actions will be called.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+    }
+
+    private ActionListener<String> internalBatchTaskListener(ADTask adTask) {
+        String taskId = adTask.getTaskId();
+        ActionListener<String> listener = ActionListener.wrap(response -> {
+            // If batch task finished normally, remove task from cache and decrease executing task count by 1.
+            adTaskCacheManager.remove(taskId);
+            adStats.getStat(AD_EXECUTING_BATCH_TASK_COUNT.getName()).decrement();
+        }, e -> {
+            // If batch task failed, remove task from cache and decrease executing task count by 1.
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:332]: Exception will be logged in `adTaskManager.handleADTaskException(adTask, e)`. 
To make debug easier, I will add a debug level log here.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+                                dataEndTime = dataEndTime - dataEndTime % interval;
+                                if ((dataEndTime - dataStartTime) < THRESHOLD_MODEL_TRAINING_SIZE * interval) {
+                                    listener
+                                        .onFailure(
+                                            new AnomalyDetectionException("There is no enough data to train model").countedInStats(false)
+                                        );
+                                    return;
+                                }
+                                long expectedPieceEndTime = dataStartTime + pieceSize * interval;
+                                long firstPieceEndTime = expectedPieceEndTime > dataEndTime ? dataEndTime : expectedPieceEndTime;
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:414]: Good point, will change

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 11]
```diff
@@ -211,7 +212,7 @@ public AnomalyDetector(
         this.description = description;
         this.timeField = timeField;
         this.indices = indices;
-        this.featureAttributes = features;
+        this.featureAttributes = features == null ? ImmutableList.of() : ImmutableList.copyOf(features);
```
Comment (yizheliu-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:215]: minor: just want to double confirm if [this line](https://github.com/opendistro-for-elasticsearch/anomaly-detection/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java#L235) will make `featureAttributes ` null?

**Review by yizheliu-amazon — APPROVED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 11 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:215]: If `featureAttributes ` is null, that line will throw exception. Actually, we can't even write such detector to stream, will throw null pointer exception.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 12]
```diff
@@ -0,0 +1,736 @@
...
+                            ADBatchTaskRemoteExecutionAction.NAME,
+                            new ADBatchAnomalyResultRequest(adTask),
+                            option,
+                            new ActionListenerResponseHandler<>(delegatedListener, ADBatchAnomalyResultResponse::new)
+                        );
+                }
+            }, e -> delegatedListener.onFailure(e))), e -> delegatedListener.onFailure(e)));
+    }
+
+    private void getNodeStats(ADTask adTask, ActionListener<DiscoveryNode> listener) {
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:222]: This method is to find the right execution node with node stats. it's better to change the method name.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 13]
```diff
@@ -0,0 +1,736 @@
...
+            }
+            candidateNodeResponse = candidateNodeResponse
+                .stream()
+                .sorted(
+                    (ADStatsNodeResponse r1, ADStatsNodeResponse r2) -> ((Long) r1
+                        .getStatsMap()
+                        .get(AD_EXECUTING_BATCH_TASK_COUNT.getName()))
+                            .compareTo((Long) r2.getStatsMap().get(AD_EXECUTING_BATCH_TASK_COUNT.getName()))
+                )
+                .collect(Collectors.toList());
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:262]: minor: in theory, this sorting isn't needed :). still ok with it since this list size won't be very large

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+                listener.onResponse(candidateNodeResponse.get(0).getNode());
+            } else {
+                // if multiple nodes have same running task count, choose the one with least JVM heap usage.
+                Long minTaskCount = (Long) candidateNodeResponse.get(0).getStatsMap().get(AD_EXECUTING_BATCH_TASK_COUNT.getName());
+                Optional<ADStatsNodeResponse> first = candidateNodeResponse
+                    .stream()
+                    .filter(c -> minTaskCount.equals(c.getStatsMap().get(AD_EXECUTING_BATCH_TASK_COUNT.getName())))
+                    .sorted(
+                        (ADStatsNodeResponse r1, ADStatsNodeResponse r2) -> ((Long) r1.getStatsMap().get(JVM_HEAP_USAGE.getName()))
+                            .compareTo((Long) r2.getStatsMap().get(JVM_HEAP_USAGE.getName()))
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:274]: we  chan change the comparator in the first sorting to avoid this filter and sorting

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+        // track AD executing batch task and total batch task execution count
+        adStats.getStat(AD_EXECUTING_BATCH_TASK_COUNT.getName()).increment();
+        adStats.getStat(StatNames.AD_TOTAL_BATCH_TASK_EXECUTION_COUNT.getName()).increment();
+
+        // put AD task into cache
+        adTaskCacheManager.put(adTask);
+
+        // start to run first piece
+        Instant executeStartTime = Instant.now();
+        runFirstPiece(adTask, executeStartTime, internalListener);
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:350]: Is there a way to make the workflow more clear? I think most of people will have the question of how the other pieces are run since the internalListener will check the whole task status. To get the idea, have to dive deep into the codes. Actually It's  in the getFeatureData method, which is hard to get it with the method name.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 16]
```diff
@@ -0,0 +1,736 @@
...
+                    }
+                } else {
+                    grade = threshold.grade(score);
+                    confidence = threshold.confidence();
+                    if (score > 0) {
+                        threshold.update(score);
+                    }
+                }
+
+                AnomalyResult anomalyResult = new AnomalyResult(
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:606]: here,  all of anomaly results will be saved, right?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 17]
```diff
@@ -0,0 +1,736 @@
...
+            long expectedPieceEndTime = pieceStartTime + pieceSize * interval;
+            long pieceEndTime = expectedPieceEndTime > dataEndTime ? dataEndTime : expectedPieceEndTime;
+            int i = 0;
+            while (i < pieceIntervalSeconds) {
+                // check if task cancelled every second, so frontend can get STOPPED state
+                // in 1 second once task cancelled.
+                checkIfADTaskCancelled(taskId);
+                rateLimiter.acquire(1);
+                i++;
+            }
```
Comment (weicongs-amazon) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:665]: why is this needed for next piece running?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 18]
```diff
@@ -699,4 +699,62 @@ public int getShingleSize(String detectorId) {
...
+
+    public void getFeatureDataPointsByBatch(
+        AnomalyDetector detector,
+        long startTime,
+        long endTime,
+        ActionListener<Map<Long, Optional<double[]>>> listener
+    ) {
+        try {
+            searchFeatureDao.getFeaturesForPeriodByBatch(detector, startTime, endTime, ActionListener.wrap(points -> {
+                logger.info("features size: {}", points.size());
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:711]: oh, yeah, will change to debug level

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 19]
```diff
@@ -699,4 +699,62 @@ public int getShingleSize(String detectorId) {
...
+                listener.onResponse(points);
+            }, listener::onFailure));
+        } catch (Exception e) {
+            logger.error("Failed to get features for detector: " + detector.getDetectorId());
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:715]: Good catch. We can have 2 options for this exception: 
Option1. It's possible that there are some bad data during some period, so we just ignore this piece and go to the next piece. So we can detect other good pieces, but user may not be able to know this error.  We will return this error and log in task `error` field. Then go to the next piece. There may be many failed pieces, we should limit max errors we can store.

Option2. Just stop the historical detector run if current piece fails. 

The option2 is easier and take less effort, but not good to tolerate bad data. Option 1 is better to handle bad data and always do the best to detect all good data. Will use option 1 to address your comments. Any suggestion?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 20]
```diff
@@ -699,4 +699,62 @@ public int getShingleSize(String detectorId) {
...
+        long maxTimeDifference = detector.getDetectorIntervalInMilliseconds() / 2;
+        Map<Long, Entry<Long, Optional<double[]>>> featuresMap = getNearbyPointsForShingle(detector, shingle, endTime, maxTimeDifference)
+            .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+        List<Entry<Long, Long>> missingRanges = getMissingRangesInShingle(detector, featuresMap, endTime);
+        missingRanges.stream().forEach(r -> {
+            if (dataPoints.containsKey(r.getKey())) {
+                featuresMap.put(r.getValue(), new SimpleImmutableEntry<>(r.getValue(), dataPoints.get(r.getKey())));
+            }
+        });
+        shingle.clear();
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:734]: The input shingle is cached and will be reused at other places like the task profile API will read this cached shingle to get realtime shingle size.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 12 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:222]: Sure, how about rename it as "dispatchTask" ?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 17 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:665]: If user stop historical detector (stop detector change will be in next PR), we will set its AD task in cache as cancelled. These lines of code is to check if AD task cancelled or not, if the task is cancelled, we will not run the next piece as `checkIfADTaskCancelled ` will throw exception. 

The piece interval is 5seconds by default, but user can tune it to a bigger value, like 60seconds, we don't want user to wait for 60seconds on AD Kibana after clicking the stop button. So this while block will check task cancelled or not every second to make sure we can set the task as cancelled in index as early as possible.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 16 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:606]: Yes, will save all anomaly results

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java [hunk 21]
```diff
@@ -0,0 +1,718 @@
...
+                    );
+                }
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
+        }, exception -> {
+            logger.error("Fail to execute onFeatureResponseLocalRCF", exception);
+            listener.onFailure(exception);
+        });
+        ThreadedActionListener threadedActionListener = new ThreadedActionListener<>(
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:524]: you miss generic type for ThreadedActionListener

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java [hunk 22]
```diff
@@ -532,14 +576,21 @@ protected Clock getClock() {
...
+            .of(
                 new FixedExecutorBuilder(
                     settings,
                     AD_THREAD_POOL_NAME,
                     Math.max(1, EsExecutors.allocatedProcessors(settings) / 4),
                     AnomalyDetectorSettings.AD_THEAD_POOL_QUEUE_SIZE,
-                    "opendistro.ad." + AD_THREAD_POOL_NAME
+                    AD_THREAD_POOL_PREFIX + AD_THREAD_POOL_NAME
+                ),
+                new FixedExecutorBuilder(
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java:588]: Can we use ScalingExecutor with minimum zero cores since not everyone uses historical analysis?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java [hunk 23]
```diff
@@ -594,7 +603,68 @@ public static boolean checkFilterByBackendRoles(User requestedUser, ActionListen
...
+
+    /**
+     * Generate batch query request for feature aggregation on given date range.
+     *
+     * @param detector anomaly detector
+     * @param startTime start time
+     * @param endTime end time
+     * @param xContentRegistry content registry
+     * @return search source builder
+     * @throws IOException throw IO exception if fail to parse feature aggregation
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:618]: Could you add AnomalyDetectionException to the comment as well?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 24]
```diff
@@ -699,4 +699,63 @@ public int getShingleSize(String detectorId) {
...
+    }
+
+    public SinglePointFeatures getShingledFeature(
+        AnomalyDetector detector,
+        Deque<Entry<Long, Optional<double[]>>> shingle,
+        Map<Long, Optional<double[]>> dataPoints,
+        long endTime
+    ) {
+        long maxTimeDifference = detector.getDetectorIntervalInMilliseconds() / 2;
+        Map<Long, Entry<Long, Optional<double[]>>> featuresMap = getNearbyPointsForShingle(detector, shingle, endTime, maxTimeDifference)
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:726]: Why do we need to call getNearbyPointsForShingle which helps real time detectors to deal with uneven arrival of requests?  You run in batches and your timestamp within an interval is fixed.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 25]
```diff
@@ -699,4 +699,63 @@ public int getShingleSize(String detectorId) {
...
+            .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+        List<Entry<Long, Long>> missingRanges = getMissingRangesInShingle(detector, featuresMap, endTime);
```
Comment (kaituo) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:728]: not sure if the workflow is needed.  You have dataPoints and shingle.  For each interval, you deque once and enqueue once.  Then you call getProcessedFeatures to interpolate.

**Review by kaituo — COMMENTED** [review]
Only finished one class FeatureManager.  Will continue reviewing.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 15 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:350]: We need to put different code logic in listener callback methods to achieve nonblocking call. So there will be multiple nested layers. There is a logical loop here: run first -> piece -> get feature data -> detect anomaly -> run next piece -> get feature data -> detector anomaly -> ...-> all pieces end. As we discussed, will add some TODO here and refactor later.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java (hunk 22 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java:588]: Good point, will change

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java (hunk 23 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/util/ParseUtils.java:618]: sure

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 24 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:726]: This is to handle sparse data, this method is used for imputing missing points in the shingle with neighboring points

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 21 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:524]: sure, will add it

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java [hunk 26]
```diff
@@ -699,4 +699,63 @@ public int getShingleSize(String detectorId) {
...
+
+        shingle.clear();
+        getFullShingleEndTimes(endTime, detector.getDetectorIntervalInMilliseconds(), detector.getShingleSize())
+            .mapToObj(time -> featuresMap.getOrDefault(time, new SimpleImmutableEntry<>(time, Optional.empty())))
+            .forEach(e -> shingle.add(e));
+
+        return getProcessedFeatures(shingle, detector, endTime);
+    }
+
+    private SinglePointFeatures getProcessedFeatures(
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:743]: Yeah, correct. If reuse existing `getProcessedFeatures` method, need to pass in listener.  Actually, we can refactor current realtime detector code to call this new method and remove the old one, no need to pass listener around here. I will add a TODO currently and do the refactor together (as you can see, weicong has some suggestion to refactor too), to avoid bringing in more and more unrelated changed to this PR which is already big enough. BTW, sorry about the big PR.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 14 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:274]: Sure will change the first comparator to include this one

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java (hunk 13 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java:262]: address with another comment https://github.com/opendistro-for-elasticsearch/anomaly-detection/pull/355#discussion_r553010777

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:725]: Only two places are using the old method, changed to new method, build can pass.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 25 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:710]: Not quite get you. Can you elaborate?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java (hunk 22 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/AnomalyDetectorPlugin.java:589]: right, ScalingExecutor is more suitable for realtime detectors.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 24 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:708]: In getNearbyPointsForShingle, the imputing distance is half of the interval, which does not apply to your case.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 25 above)
Comment (kaituo) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:710]: I meant your workflow can be:
1) you know the next timestamp t
2) check if dataPoints has a timestamp t
3) if yes, deque the shingle and enque the data point at 2)
4) if no, try interpolation.

The realtime part of the code is hard to understand and not necessary for your use case.

**Review by weicongs-amazon — APPROVED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 24 above)
Comment (weicongs-amazon) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:708]: Do we really need this logic for historical detector?

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:708]: This is from product/user experience consideration, we will not differentiate the historical and realtime detection, will keep the model/algorithm consistent.

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:708]: We don't need to impute missing data which caused by run time jitter for historical detector, just need to impute the data hole in source data. 

Discussed with kaituo, will simplify the code for historical detector currently. For the single flow, we can create new function to handle both historical and realtime detection.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 25 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:710]: This comment discusses the same issue with this [one](https://github.com/opendistro-for-elasticsearch/anomaly-detection/pull/355/files#r553062489).

This function was designed to handle both realtime and historical detector. It looks complicated. Will simplify this function for historical detector. May create new function for single flow later.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java (hunk 24 above)
Comment (weicongs-amazon) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java:708]: seems that we have lots of ideas for the new universal workflow. Should we create a doc/issue to track them?

Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/feature/FeatureManager.java]: Yeah, as universal flow is just started, I will doc these ideas on my notebook only now. Don't thinks it will benefit community to share such unconnected ideas on Github now as we don't even mention what universal flow is. Will create an RFC issue later once we finish research, and put those ideas to that Github issue. So user can know the big picture of background, our solution, then they can understand why we come up with these ideas.

**Comment by kaituo** [comment]
Thanks for the change for FeatureManager, Yaliang.  I approve the change to that file.
